city_scrapers/spiders/il_pollution_control.py

```diff
@@ -0,0 +1,160 @@
+import json
+from datetime import datetime
+from urllib.parse import urljoin
+
+import scrapy
+from city_scrapers_core.constants import BOARD, NOT_CLASSIFIED
+from city_scrapers_core.items import Meeting
+from city_scrapers_core.spiders import CityScrapersSpider
+
+
+class IlPollutionControlSpider(CityScrapersSpider):
+    name = "il_pollution_control"
+    agency = "Illinois Pollution Control Board"
+    timezone = "America/Chicago"
+    allowed_domains = ["pcb.illinois.gov"]
+    start_urls = ["https://pcb.illinois.gov/ClerksOffice/MeetingMinutes"]
+    json_url = "https://pcb.illinois.gov/ClerksOffice/GetCalendarEvents"
+    calendar_url = "https://pcb.illinois.gov/ClerksOffice/Calendar"
+
+    def __init__(self, *args, **kwargs):
+        self.link_map = dict()  # Populated by self._parse_minutes()
+        self.relevant_years = [
+            str(y) for y in range(datetime.now().year - 1,
+                                  datetime.now().year + 1)
+        ]
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def from_crawler(cls, crawler, *args, **kwargs):
+        """ Overridden `from_crawler` to connect `spider_idle` signal. """
+        spider = super().from_crawler(crawler, *args, **kwargs)
+        crawler.signals.connect(spider.spider_idle, signal=scrapy.signals.spider_idle)
+        return spider
+
+    def spider_idle(self):
+        """ React to `spider_idle` signal by starting JSON parsing after _parse_minutes."""
+        self.crawler.signals.disconnect(self.spider_idle, signal=scrapy.signals.spider_idle)
+        self.crawler.engine.crawl(scrapy.Request(self.json_url, callback=self._parse_json), self)
+        raise scrapy.exceptions.DontCloseSpider
+
+    def parse(self, response):
+        """
+        `parse` should always `yield` Meeting items.
+        """
+        for item in response.xpath("//iframe/@src"):
+            yield scrapy.Request(item.get(), callback=self._parse_minutes)
+
+    def _parse_minutes(self, response):
+        """ Traverse tree of URLs and populate self.link_map """
+        for item in response.xpath("//td[@class='name']/a"):
+            try:
+                href = item.xpath("@href")[0].get()
+                text = item.xpath("b/text()")[0].get().strip()
+                if not any([(year in text) for year in self.relevant_years]):
+                    continue  # Link does not contain documents from recent years
+                if text[-4:] == ".pdf":
+                    text = text[:-4]
+            except IndexError:
+                continue
+
+            url = urljoin("https://{}".format(self.allowed_domains[0]), href)
```

Comment:
Missed this earlier, but looking at it now from the tests and unless I'm missing something `response.urljoin(href)` should be quicker here

Reply:
I seem to remember there being a problem with that approach but I'm not 100% sure there; I'll take a look and verify.